packages/flutter/lib/src/widgets/animated_grid.dart

```diff
@@ -0,0 +1,714 @@
+          'The context used was:\n'
+          '  $context',
+        );
+      }
+      return true;
+    }());
+    return result!;
+  }
+
+  /// The state from the closest instance of this class that encloses the given
+  /// context.
+  ///
+  /// This method is typically used by [SliverAnimatedGrid] item widgets that
+  /// insert or remove items in response to user input.
+  ///
+  /// If no [SliverAnimatedGrid] surrounds the context given, then this function
+  /// will return null.
+  ///
+  /// This method can be expensive (it walks the element tree).
+  ///
+  /// See also:
+  ///
+  ///  * [of], a similar function that will throw if no [SliverAnimatedGrid]
+  ///    ancestor is found.
+  static SliverAnimatedGridState? maybeOf(BuildContext context) {
+    assert(context != null);
+    return context.findAncestorStateOfType<SliverAnimatedGridState>();
+  }
+}
+
+/// The state for a sliver that animates items when they are
+/// inserted or removed.
+///
+/// When an item is inserted with [insertItem] an animation begins running. The
+/// animation is passed to [SliverAnimatedGrid.itemBuilder] whenever the item's
+/// widget is needed.
+///
+/// When an item is removed with [removeItem] its animation is reversed.
+/// The removed item's animation is passed to the [removeItem] builder
+/// parameter.
+///
+/// An app that needs to insert or remove items in response to an event
+/// can refer to the [SliverAnimatedGrid]'s state with a global key:
+///
+/// ```dart
+/// // (e.g. in a stateful widget)
+/// GlobalKey<AnimatedGridState> gridKey = GlobalKey<AnimatedGridState>();
+///
+/// // ...
+///
+/// @override
+/// Widget build(BuildContext context) {
+///   return AnimatedGrid(
+///     key: gridKey,
+///     itemBuilder: (BuildContext context, int index, Animation<double> animation) {
+///       return const Placeholder();
+///     },
+///     gridDelegate: const SliverGridDelegateWithMaxCrossAxisExtent(maxCrossAxisExtent: 100.0),
+///   );
+/// }
+///
+/// // ...
+///
+/// void _updateGrid() {
+///   // adds "123" to the AnimatedGrid
+///   gridKey.currentState!.insertItem(123);
+/// }
+/// ```
+///
+/// [SliverAnimatedGrid] item input handlers can also refer to their
+/// [SliverAnimatedGridState] with the static [SliverAnimatedGrid.of] method.
+class SliverAnimatedGridState extends State<SliverAnimatedGrid> with TickerProviderStateMixin {
+  final List<_ActiveItem> _incomingItems = <_ActiveItem>[];
+  final List<_ActiveItem> _outgoingItems = <_ActiveItem>[];
+  int _itemsCount = 0;
+
+  @override
+  void initState() {
+    super.initState();
+    _itemsCount = widget.initialItemCount;
+  }
+
+  @override
+  void dispose() {
+    for (final _ActiveItem item in _incomingItems.followedBy(_outgoingItems)) {
+      item.controller!.dispose();
+    }
+    super.dispose();
+  }
+
+  _ActiveItem? _removeActiveItemAt(List<_ActiveItem> items, int itemIndex) {
```

Comment:
Reading through this and comparing to SliverAnimatedList, it might be better to refactor so that AnimatedList/Grid are in one file, and SliverAnimatedList/Grid are in another. That way a lot of these methods can be shared between them in like a mixin. The _ActiveItem class could be de-duped then too.

Reply:
I can do this refactor in a follow up change if you want. :)

Reply:
Confirmed offline, I'll refactor after this lands. This PR is already big enough.